### PR TITLE
[MRG] Added tests for the function `load_pathlist_from_file`

### LINKED ISSUE #1430: Add tests in `tests/test_sourmash.py` for the function `load_pathlist_from_file`
#1423 partly fixes issue #1369; this issue is for the remaining bit of #1369.

Please add some Python unit tests in ```tests/test_sourmash.py``` for the ```load_pathlist_from_file``` function, including empty file list, badly formatted ones, ones with duplicates.

### PR DESCRIPTION
**Pull request by keyabarve**
Fixes #1430

### CODE REVIEW

**Comment by keyabarve** [comment]
I wrote 2 tests for Badly Formatted Files and 1 test for Duplicate Files. I came across some interesting errors that need to be worked on.
@luizirber

**Comment by ctb** [comment]
hi @keyabarve could you summarize the errors here in the comments, and (if you're up to it) propose some solutions? thanks!

**Comment by keyabarve** [comment]
@ctb @luizirber 
Tests are written in `tests/test_sourmash.py`. The following are the different issues in each of the test cases that need to be fixed. I'm not entirely sure why these issues are arising and how to fix them.

- Function `test_load_pathlist_from_file_empty`: 
This function is testing if the given file is empty. The file is written with no contents ("").
An IndexError is raised, stating "list index out of range".

- Function `test_load_pathlist_from_file_badly_formatted`: 
The function is testing if the given file is badly formatted. A bad input ("{'a':1}") is written into the file.
ValueError is raised, stating "first element of list-of-files does not exist".

- Function `test_load_pathlist_from_file_badly_formatted_2`: 
The function is testing if the second line of the file is badly formatted, given that the first line of the file is valid. The contents of an existing file ("compare/genome-s10.fa.gz.sig") are copied into the first line of the file and a bad input ("{'a':1}") is written in the second line.
`assert len(sigs)` returns a value of 2.

- Function `test_load_pathlist_from_file_duplicate`: 
The function is testing if the same file is passed more than once. The contents of an existing file ("compare/genome-s10.fa.gz.sig") are copied into the first line as well as the second line of the file.
`assert len(sigs)` returns a value of 2.

File: tests/test_sourmash.py [hunk 1]
```diff
@@ -72,6 +72,74 @@ def test_sourmash_info_verbose():
     assert "loaded from path" in err
 
 
+def test_load_pathlist_from_file_does_not_exist():
+    from sourmash.sourmash_args import load_pathlist_from_file
+    with pytest.raises(ValueError) as e:
+        load_pathlist_from_file("")
+        assert "cannot open file ''" in e.message
```
Comment (ctb) [inline tests/test_sourmash.py]: note, you should de-indent here, because the `assert` statement is never reached - this is because `load_pathlist_from_file` raises a ValueError, which ends the block and is caught by the `with` statement.

**Comment by ctb** [comment]
thanks! These are good checks, and I agree `load_pathlist_from_file` should return better errors on them!

To fix them, you'd want to have `load_pathlist_from_file` raise `ValueError` with appropriate error messages for each of them.

For example, to fix the `empty` error, you could add another `if` statement  in https://github.com/dib-lab/sourmash/blob/beb3d5963332aba162c530cf612739206d7efbd6/src/sourmash/sourmash_args.py#L389 that checks to see if `file_list` is empty or not, and if it is, raises a `ValueError`.

Is this something you want to pursue?

**Comment by keyabarve** [comment]
Yes, I would like to continue working on this. I will try fixing it.

File: tests/test_sourmash.py (hunk 1 above)
Comment (keyabarve) [reply tests/test_sourmash.py]: Which lines should I de-indent?

Comment (ctb) [reply tests/test_sourmash.py]: I'll ask a question instead ;).

try putting an `assert 0` after the `load_pathlist_from_file` call - does the test fail, as you'd expect it would?

also see:

- [asserting with the `assert` statement](https://docs.pytest.org/en/stable/assert.html#asserting-with-the-assert-statement)
- [`with pytest.raises` documentation](https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions)
- [context managers](https://www.python.org/dev/peps/pep-0343/#id36

[python exceptions and try/except blocks](https://www.w3schools.com/python/python_try_except.asp) may also be useful.

And note that you can put `print` statements in tests and the output will be displayed only when the test fails, so they're useful for debugging tests.

**Comment by ctb** [comment]
> Yes, I would like to continue working on this. I will try fixing it.

excellent!

I think a good high level summary of what to do overall is "take any of these conditions or errors and turn them into statements that raise `ValueError` (with informative error messages) in the `load_pathlist_from_file` function."

**Comment by keyabarve** [comment]
@ctb @luizirber 
I had a few questions regarding the function `load_pathlist_from_file`:
- What exactly does the argument "filename" passed to the function `load_pathlist_from_file` contain? Is "filename" a single file, which contains a list of files? 
- What does the variable "file_list" that is returned by this function contain? Is it an array containing the list of files from "filename"?
- When we are checking for all the test cases (file doesn't exist, empty file, badly formatted file, and duplicate file), are we checking all these for each of the files inside "filename"?
- So, while writing the different if statements in this function, should I loop through "file_list" to check each of the files?

**Comment by ctb** [comment]
On Mon, Apr 26, 2021 at 05:17:55PM -0700, Keya Barve wrote:
> - What exactly does the argument "filename" passed to the function `load_pathlist_from_file` contain? Is "filename" a single file, which contains a list of files? 

yes.

> - What does the variable "file_list" that is returned by this function contain? Is it an array containing the list of files from "filename"?

that is defined by the function, right? there's no mystery here, you have
the full source code :).

(what is the answer, based on the current code?)

> - When we are checking for all the test cases (file doesn't exist, empty file, badly formatted file, and duplicate file), are we checking all these for each of the files inside "filename"?

I believe that most of those conditions are handled by existing test code, so
you don't need to check them explicitly.

The only exception is duplicate file. I think perhaps we should collapse
duplicate entries in the file by returning a set rather than a list.

> - So, while writing the different if statements in this function, should I loop through "file_list" to check each of the files?

Nope.

You might be interested in reading up on "duck typing" in Python, BTW;
as it is, the load_pathlist function takes any stringlike object and
returns an iterable, but those are the only operational constraints at
the moment. And that's intentional; there's no reason to be more specific
about what this function takes and returns.

**Comment by keyabarve** [comment]
Alright, thanks!

**Comment by luizirber** [comment]
Related discussion: #1410

**Comment by keyabarve** [comment]
@ctb @luizirber 
Whenever I test any of the test cases, it gives me a type error, saying ```TypeError: stat: path should be string, bytes, os.PathLike or integer, not list``` 
This error is coming from the function: 
```
def exists(path):
        """Test whether a path exists.  Returns False for broken symbolic links"""
        try:
>           os.stat(path)
E           TypeError: stat: path should be string, bytes, os.PathLike or integer, not list
```
Could you provide some tips on how I can fix this?

**Comment by ctb** [comment]
On Fri, Apr 30, 2021 at 11:24:24AM -0700, Keya Barve wrote:
> @ctb @luizirber 
> Whenever I test any of the test cases, it gives me a type error, saying ```TypeError: stat: path should be string, bytes, os.PathLike or integer, not list``` 
> This error is coming from the function: 
> ```def exists(path):
>         """Test whether a path exists.  Returns False for broken symbolic links"""
>         try:
> >           os.stat(path)
> E           TypeError: stat: path should be string, bytes, os.PathLike or integer, not list```


what is path?

do:

print((path,), type(path))

**Comment by keyabarve** [comment]
Where should I write this print statement? Should I write it in `load_pathlist_from_file`? There is no path variable in this function, so I'm not sure if it will print anything.

**Comment by ctb** [comment]
On Fri, Apr 30, 2021 at 11:28:34AM -0700, Keya Barve wrote:
> Where should I write this print statement? Should I write it in `load_pathlist_from_file`? There is no path variable in this function, so I'm not sure if it will print anything.

either where you call load_pathlist_from_file, or in load_pathlist from file.

please remember you control all the code here, so you can print out variables
where you want :)

**Comment by keyabarve** [comment]
@ctb @luizirber 
I think the code works fine for all the test cases except for the duplicate files. Since I have created a loop, should I remove the `test_load_pathlist_from_file_badly_formatted_2` test case?

**Comment by ctb** [comment]
> @ctb @luizirber
> I think the code works fine for all the test cases except for the duplicate files. Since I have created a loop, should I remove the `test_load_pathlist_from_file_badly_formatted_2` test case?

what about collapsing `file_list` automatically, using an [ordered dictionary](https://docs.python.org/3/library/collections.html#collections.OrderedDict) so that the files remain in the same order but get collapsed? You could also do that within your for loop using a `set` underneath to see if we've already seen a file.

We probably don't need to emit a warning in the case of duplicate files, because if you passed duplicate files into most sourmash commands, sourmash would invisibly deduplicate them anyway in search/gather/sketch).

**Comment by keyabarve** [comment]
> > @ctb @luizirber
> > I think the code works fine for all the test cases except for the duplicate files. Since I have created a loop, should I remove the `test_load_pathlist_from_file_badly_formatted_2` test case?
> 
> what about collapsing `file_list` automatically, using an [ordered dictionary](https://docs.python.org/3/library/collections.html#collections.OrderedDict) so that the files remain in the same order but get collapsed? You could also do that within your for loop using a `set` underneath to see if we've already seen a file.
> 
> We probably don't need to emit a warning in the case of duplicate files, because if you passed duplicate files into most sourmash commands, sourmash would invisibly deduplicate them anyway in search/gather/sketch).

What do you mean by collapsing `file_list`? Should I be doing this for the duplicate files?

Also, should I remove the test case for the duplicate files entirely?

**Comment by ctb** [comment]
On Fri, Apr 30, 2021 at 12:58:09PM -0700, Keya Barve wrote:
> > > @ctb @luizirber
> > > I think the code works fine for all the test cases except for the duplicate files. Since I have created a loop, should I remove the `test_load_pathlist_from_file_badly_formatted_2` test case?
> > 
> > what about collapsing `file_list` automatically, using an [ordered dictionary](https://docs.python.org/3/library/collections.html#collections.OrderedDict) so that the files remain in the same order but get collapsed? You could also do that within your for loop using a `set` underneath to see if we've already seen a file.
> > 
> > We probably don't need to emit a warning in the case of duplicate files, because if you passed duplicate files into most sourmash commands, sourmash would invisibly deduplicate them anyway in search/gather/sketch).
> 
> What do you mean by collapsing `file_list`? Should I be doing this for the duplicate files?

collapsing => ignoring duplicates.

> Also, should I remove the test case for the duplicate files entirely?

I would only remove tests if they are entirely redundant or serve
no purpose whatsoever.

File: src/sourmash/sourmash_args.py [hunk 2]
```diff
@@ -391,14 +391,22 @@ def load_pathlist_from_file(filename):
     try:
         with open(filename, 'rt') as fp:
             file_list = [ x.rstrip('\r\n') for x in fp ]
-
-        if not os.path.exists(file_list[0]):
-            raise ValueError("first element of list-of-files does not exist")
+        if len(file_list) == 0:
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: note, this can be abbreviated more Pythonically as:
```
if not file list:
...
```
(in Python, empty lists evaluate to False)

Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Okay, I'll fix that!

File: src/sourmash/sourmash_args.py [hunk 3]
```diff
@@ -391,14 +391,22 @@ def load_pathlist_from_file(filename):
...
+            raise ValueError("list-of-files is empty")
+        cnt = 0
+        for i in range(len(file_list)):
+            if not os.path.exists(file_list[i]):
+                cnt = cnt + 1
+        if cnt > 0:
+            raise ValueError("list-of-files contains a badly formatted file")
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: Is this error message right?  You're checking if the file exists; if I were a user I would not know that "badly formatted" means "does not exist"

Also, suggest using "pathlist" rather than "list of files".

Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Could you suggest a way to check for badly formatted files?

File: src/sourmash/sourmash_args.py [hunk 4]
```diff
@@ -391,14 +391,22 @@ def load_pathlist_from_file(filename):
     try:
```
Comment (ctb) [inline src/sourmash/sourmash_args.py:393]: note that the try/except should probably be _just_ around the `with open...` block. Generally speaking we want to avoid catching exceptions over-broadly - if we wrap a dozen lines of code in a try/except, we won't know where the error comes from and won't be able to report the problem to the user! (This is a weak rule of thumb, there are some situations where you can't help it, of course. Just pointing it out as an option for you to consider!)

**Comment by ctb** [comment]
> Could you suggest a way to check for badly formatted files?

I'm afraid I don't really have any good suggestions off the top of my head!

File: src/sourmash/sourmash_args.py (hunk 4 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py:393]: Ahh I see... Where can I put the if statements instead?

**Comment by keyabarve** [comment]
@ctb @luizirber 
This is the latest version of my code. According to this, all the test cases pass. I still need to incorporate some of the changes suggested.

File: src/sourmash/sourmash_args.py (hunk 4 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py:393]: outside the try/except block should work? That may not work for other reasons, but the way scoping works in Python `file_list` will be available outside of the try/except block, if I'm intuiting your concern correctly.

**Comment by ctb** [comment]
cool - let us know when it's ready to review!

you should also remember to update the branch from `latest` at some point, since other changes have been merged into that branch since you started work. You can do that with `git pull origin latest; git push` or you can select the `update branch` button on the pull request and then do a `git pull` on your machine so that you have the latest changes locally, too.

**Comment by ctb** [comment]
hi @keyabarve please see the code in https://github.com/dib-lab/sourmash/pull/1492 - your tests aren't actually testing the things you think they are ;). Ask questions here or there, happy to help.

**Comment by keyabarve** [comment]
@ctb
I have de-indented the assert statements and all the test cases pass.

**Comment by keyabarve** [comment]
@ctb Does the code look fine now? I'm still failing one of the checks though.

**Comment by keyabarve** [comment]
Ready for review and merge. @ctb

File: src/sourmash/sourmash_args.py (hunk 2 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py]: bump!

File: src/sourmash/sourmash_args.py [hunk 5]
```diff
@@ -393,14 +393,27 @@ def load_pathlist_from_file(filename):
...
         with open(filename, 'rt') as fp:
             file_list = [ x.rstrip('\r\n') for x in fp ]
-
-        if not os.path.exists(file_list[0]):
-            raise ValueError("first element of list-of-files does not exist")
+        if len(file_list) == 0:
+            raise ValueError("pathlist is empty")
+        badFormat = 0
+        duplicate = []
+        for i in range(len(file_list)):
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: a few comments -

* duplicate should be a `set`, not a `list`; lists are linear time traversal, sets are random lookups based on hash value.
* `badformat` could be better used as a boolean, so `badFormat = False`;
* then you can use `if badFormat`;
* the for loop can also be `for checkfile in file_list`, no need to use range;

File: src/sourmash/sourmash_args.py [hunk 6]
```diff
@@ -393,14 +393,27 @@ def load_pathlist_from_file(filename):
...
+            if not os.path.exists(file_list[i]):
+                badFormat = badFormat + 1
+            if file_list[i] not in duplicate:
+                duplicate.append(file_list[i])
+        if badFormat > 0:
+            raise ValueError("pathlist contains a badly formatted file")
+        if len(duplicate) != len(file_list):
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: see [this comment](https://github.com/dib-lab/sourmash/pull/1469#issuecomment-830353364). I don't think we need to warn about duplicates, we should just remove them (by making `file_list` a set).

File: tests/test_sourmash.py [hunk 7]
```diff
@@ -72,6 +72,76 @@ def test_sourmash_info_verbose():
     assert "loaded from path" in err
 
 
+def test_load_pathlist_from_file_does_not_exist():
+    from sourmash.sourmash_args import load_pathlist_from_file
+    with pytest.raises(ValueError) as e:
+        load_pathlist_from_file("")
+    assert "file '' does not exist" in str(e.value)
+
```
Comment (ctb) [inline tests/test_sourmash.py:81]: please add two lines between functions, thanks!

**Review by ctb — CHANGES_REQUESTED** [review]
Hi! the tests look great, but I requested some changes to the implementation. Thanks!

File: tests/test_sourmash.py (hunk 7 above)
Comment (keyabarve) [reply tests/test_sourmash.py:81]: Okay, I'll do that!

File: src/sourmash/sourmash_args.py (hunk 6 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: So basically I created a set called `duplicate` and I added each item in `file_list` to it. Since it is a set, it will not add any duplicate items. Is this not how I should do it?

File: src/sourmash/sourmash_args.py (hunk 5 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Okay, I'll fix all of that!

File: src/sourmash/sourmash_args.py (hunk 6 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py]: err, here `duplicate` is a list, is it not?

Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Oh I think I messed it up! Will fix it, thanks! Just to clarify, should I do the same thing I'm currently doing after making `duplicate` a set?

Comment (ctb) [reply src/sourmash/sourmash_args.py]: what a great question! submit something that passes the tests and let me know when you want me to review it ;)

**Comment by keyabarve** [comment]
Please review. @ctb

**Comment by ctb** [comment]
Note: changing title to [MRG] since it's ready for review.

File: src/sourmash/sourmash_args.py (hunk 2 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py]: Please revisit this comment, thanks - https://github.com/dib-lab/sourmash/pull/1469/files#r624171173

File: src/sourmash/sourmash_args.py [hunk 8]
```diff
@@ -393,14 +393,27 @@ def load_pathlist_from_file(filename):
...
-        if not os.path.exists(file_list[0]):
-            raise ValueError("first element of list-of-files does not exist")
+        if len(file_list) == 0:
+            raise ValueError("pathlist is empty")
+        badFormat = False
+        duplicate = set()
+        for checkfile in file_list:
+            if not os.path.exists(checkfile):
+                badFormat = True
+            if checkfile not in duplicate:
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: This `if` is unnecessary! You can just say `duplicate.add(checkfile)` - it's actually faster, in addition to being more concise!

File: src/sourmash/sourmash_args.py [hunk 9]
```diff
@@ -393,14 +393,27 @@ def load_pathlist_from_file(filename):
...
+            raise ValueError("pathlist contains a duplicate file")
+
+    except IOError:
+        raise ValueError(f"file '{filename}' does not exist")    
     except OSError:
         raise ValueError(f"cannot open file '{filename}'")
     except UnicodeDecodeError:
         raise ValueError(f"cannot parse file '{filename}' as list of filenames")
-
+
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: please remove whitespace other than newline :)

File: tests/test_sourmash.py [hunk 10]
```diff
@@ -72,6 +72,77 @@ def test_sourmash_info_verbose():
     assert "loaded from path" in err
 
 
+def test_load_pathlist_from_file_does_not_exist():
+    from sourmash.sourmash_args import load_pathlist_from_file
```
Comment (ctb) [inline tests/test_sourmash.py:77]: pls move import statement to top of file and remove from all of the functions. No need to run it each time when you can do it at the module level!

File: tests/test_sourmash.py [hunk 11]
```diff
@@ -72,6 +72,77 @@ def test_sourmash_info_verbose():
...
+    new_file = os.path.join(newdir, 'sig1')
+    shutil.copyfile(sig1, new_file)
+
+    with open(file_list, "w") as fp:
+        fp.write(new_file + "\n")
+        fp.write(new_file + "\n")
+
+    with pytest.raises(ValueError) as e:
+        load_pathlist_from_file(file_list)
+    assert "pathlist contains a duplicate file" in str(e.value)
```
Comment (ctb) [inline tests/test_sourmash.py]: so, I still think we shouldn't care about duplicate files, and we should just remove them silently. See comment [here](https://github.com/dib-lab/sourmash/pull/1469#issuecomment-827224466) suggesting that we do that silently and return a set rather than a list from this function.

Comment (ctb) [reply tests/test_sourmash.py]: to implement this, you could make `file_list` into a set immediately the list comprehension, and remove all mention of `duplicate`.

File: tests/test_sourmash.py [hunk 12]
```diff
@@ -72,6 +72,77 @@ def test_sourmash_info_verbose():
...
+    file_list = c.output("file_list")
+
+    sig1 = utils.get_test_data('compare/genome-s10.fa.gz.sig')
+    newdir = c.output('newdir')
+    os.mkdir(newdir)
+    new_file = os.path.join(newdir, 'sig1')
+    shutil.copyfile(sig1, new_file)
+
+    with open(file_list, "w") as fp:
+        fp.write(new_file + "\n")
```
Comment (ctb) [inline tests/test_sourmash.py]: note that you don't need to copy `sig1` to `new_file`, you can put the path for `sig1` directly into the file_list file. pls change ;).

**Review by ctb — CHANGES_REQUESTED** [review]
A few more requests - thanks! A lot of these are just Pythonisms and so on that you don't know yet, so don't despair - you're getting closer!

File: src/sourmash/sourmash_args.py (hunk 9 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Yes, will do that!

File: tests/test_sourmash.py (hunk 11 above)
Comment (keyabarve) [reply tests/test_sourmash.py]: So if I convert `file_list` into a set, will it take care of duplicates on its own? Should I do this conversion in `tests/test_sourmash.py` or should I do it inside the function itself in `src/sourmash/sourmash_args.py`?

File: tests/test_sourmash.py (hunk 12 above)
Comment (keyabarve) [reply tests/test_sourmash.py]: Oh I see, I will change that.

Comment (keyabarve) [reply tests/test_sourmash.py]: Should I remove all of the following lines then?
`newdir = c.output('newdir')`
`os.mkdir(newdir)`
`new_file = os.path.join(newdir, 'sig1')`
`shutil.copyfile(sig1, new_file)`

And then, in `fp.write()` on line 138, should I put `sig1` instead of `new_file`?

File: tests/test_sourmash.py (hunk 10 above)
Comment (keyabarve) [reply tests/test_sourmash.py:77]: Sure!

File: src/sourmash/sourmash_args.py (hunk 8 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Which `if` is this? is it the one on line 401, or the one on line 403?

File: tests/test_sourmash.py (hunk 11 above)
Comment (keyabarve) [reply tests/test_sourmash.py]: Should we return a set from `src/sourmash/sourmash_args.py` instead of a list? In this case, I would have to write the code as something like this:
`with open(filename, 'rt') as fp:`
`temp_file_list = [ x.rstrip('\r\n') for x in fp ]`
`file_list = set(temp_file_list)`

I'm still not sure how to handle duplicate files. If I modify the code this way, should I not care about them at all? Should I remove all the code relating to the variable `duplicate` in `src/sourmash/sourmash_args.py`? And should I even remove the test case for the duplicate files in `tests/test_sourmash.py`?

Comment (ctb) [reply tests/test_sourmash.py]: That should all work, yes. I think it's ok to silently ignore duplicate files.

I think having the test case for duplicate files is good, but it should verify that they are ignored instead of checking for the exception.

File: tests/test_sourmash.py (hunk 12 above)
Comment (ctb) [reply tests/test_sourmash.py]: please give it a try!

File: tests/test_sourmash.py (hunk 11 above)
Comment (ctb) [reply tests/test_sourmash.py]: regarding where to do the conversion -

the code in src is the application that is distributed, and that should have the behavior of silently collapsing and ignoring duplicates.

the tests check the behavior. they are not distributed with the application.

I suggest giving it a try and then asking for a review once it works ;).

File: src/sourmash/sourmash_args.py (hunk 8 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py]: the only `if ` that involves `duplicate` is on line 403.

File: tests/test_sourmash.py (hunk 11 above)
Comment (keyabarve) [reply tests/test_sourmash.py]: How can I test if the duplicates have been ignored?

Comment (ctb) [reply tests/test_sourmash.py]: if you know how many distinct entries there are in the pathlist file, you can look at the return value of the `load_pathlist_from_file` function, no?

**Comment by keyabarve** [comment]
Please review. @ctb

**Comment by keyabarve** [comment]
I started working on some of the test cases related to another issue, in `tests/test_sourmash.py` itself and they probably got pushed with the changes related to this issue. Is that why some of the checks here are failing? Should I undo the other changes and push it again?

File: src/sourmash/sourmash_args.py [hunk 13]
```diff
@@ -392,15 +392,22 @@ def load_pathlist_from_file(filename):
     "Load a list-of-files text file."
     try:
         with open(filename, 'rt') as fp:
-            file_list = [ x.rstrip('\r\n') for x in fp ]
-
-        if not os.path.exists(file_list[0]):
-            raise ValueError("first element of list-of-files does not exist")
+            temp_file_list = [ x.rstrip('\r\n') for x in fp ]
+        file_list = set(temp_file_list)
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: suggestion - I usually prefer just using the same variable name, e.g. get rid of `temp_file_list` altogether:
```
file_list = set(file_list)
```
Because of the way Python works, the value on the left is only set after the expression on the right is evaluated (ok that's normal for many languages :) and then the value that was used on the right is dereferenced and garbage collected (that's Python!) This has the slight benefit that you don't have the `temp_file_list` variable hanging around uselessly, either.

File: tests/test_sourmash.py [hunk 14]
```diff
@@ -198,7 +250,8 @@ def test_do_basic_compare_using_rna_arg(c):
 def test_do_compare_quiet(c):
     testdata1 = utils.get_test_data('short.fa')
     testdata2 = utils.get_test_data('short2.fa')
-    c.run_sourmash('compute', '-k', '31', testdata1, testdata2)
+    # c.run_sourmash('compute', '-k', '31', testdata1, testdata2)
```
Comment (ctb) [inline tests/test_sourmash.py]: ok - this does seem out of place 😆 . would you like me to remove it or do you want me to coach you through removing it?

File: src/sourmash/sourmash_args.py (hunk 13 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: Okay! I'll fix that!

File: tests/test_sourmash.py (hunk 14 above)
Comment (keyabarve) [reply tests/test_sourmash.py]: I'd actually like to learn.. :)

File: src/sourmash/sourmash_args.py [hunk 15]
```diff
@@ -392,15 +392,22 @@ def load_pathlist_from_file(filename):
...
+        if not file_list:
+            raise ValueError("pathlist is empty")
+        badFormat = False
+        for checkfile in file_list:
+            if not os.path.exists(checkfile):
+                badFormat = True
+        if badFormat:
+            raise ValueError("pathlist contains a badly formatted file")
+    except IOError:
+        raise ValueError(f"file '{filename}' does not exist")
```
Comment (ctb) [inline src/sourmash/sourmash_args.py]: hmm, I'm confused by the error messages you chose here -

1. if the file *inside* the pathlist doesn't exist, the error message is `pathlist contains a badly formatted file`
2. if the pathlist file itself doesn't exist, the error message is "file does not exist".

A suggestion --

* if the file inside the pathlist file doesn't exist, the error message could be "file X listed inside pathlist Y does not exist"
* if the pathlist file itself doesn't exist (the IOError), the error message could be "pathlist file 'filename' doesn't exist"

File: tests/test_sourmash.py (hunk 14 above)
Comment (ctb) [reply tests/test_sourmash.py]: so, in this case, I think the simplest thing to do is change the code back in a new commit - just remove the comment on the old line, and remove the new line. Once this is merged all that history will be gone.

In the future, there are a variety of techniques you can use such as `git revert` to choose an old commit in the history to revert, if the change is the only thing in that commit; and you can use `git reset HEAD^` to go back to the previous commit (current minus one), if it's the most recent commit. I basically use these three options 99% of the time.

File: src/sourmash/sourmash_args.py (hunk 15 above)
Comment (keyabarve) [reply src/sourmash/sourmash_args.py]: In this case, should I remove `badformat` altogether? Should I just raise the error inside the for loop? I could write something like this:
`for checkfile in file_list:`
`if not os.path.exists(checkfile):`
`raise ValueError("pathlist contains a badly formatted file")`

Comment (ctb) [reply src/sourmash/sourmash_args.py]: I like it!! Yes! (note that you can do multiline things with indentation by using triple-back-quotes, so 
```
`` `
```
(but without the space ;)

**Comment by keyabarve** [comment]
@ctb I made all the changes you requested. Please review.

**Review by ctb — APPROVED** [review]
Nice work!

**Comment by ctb** [comment]
Whups! I approved this, but didn't notice that two tests were failing -

```
tests/test_cmd_signature.py::test_sig_cat_filelist_with_dbs
tests/test_cmd_signature.py::test_sig_cat_filelist
```

both depended on pathlist being a list. So they need to be fixed as well.

**Comment by keyabarve** [comment]
Hmm, I see. Could you give me some suggestions on how I can fix these tests?

**Comment by keyabarve** [comment]
I have made all the changes and pushed them. I expect it to be working properly now, but it still fails both the tests. I can't seem to figure out what the error is. 
Please review. @ctb

**Comment by ctb** [comment]
hi @keyabarve take a look at https://github.com/dib-lab/sourmash/pull/1527/files and see if that makes sense to you.

**Comment by ctb** [comment]
hi @keyabarve looks like there's still a test failing 😭 . Could you take a look and give it a stab, or ask some questions?

**Comment by keyabarve** [comment]
@ctb Sure! I'll take a look at it! Which test is failing exactly?

**Comment by ctb** [comment]
On Mon, May 17, 2021 at 06:25:53PM -0700, Keya Barve wrote:
> @ctb Sure! I'll take a look at it! Which test is failing exactly?

you can dig around in the "some checks were not successful" to find out,
or run all the tests yourself, locally.

**Comment by ctb** [comment]
hi @keyabarve you could usefully add a test to catch this situation: https://github.com/dib-lab/sourmash/issues/1537

**Comment by ctb** [comment]
sorry, a fix AND a test :)

**Comment by keyabarve** [comment]
@ctb My latest code already takes care of the issue and I already have a test for it! :)

File: tests/test_sourmash.py [hunk 16]
```diff
@@ -199,7 +252,6 @@ def test_do_compare_quiet(c):
     testdata1 = utils.get_test_data('short.fa')
     testdata2 = utils.get_test_data('short2.fa')
     c.run_sourmash('compute', '-k', '31', testdata1, testdata2)
-
```
Comment (ctb) [inline tests/test_sourmash.py:202]: note - please try to avoid changes in bits of the code that aren't relevant to the issues tackled by this PR - it adds a bit of confusion to the review :)

Comment (keyabarve) [reply tests/test_sourmash.py:202]: Sorry, it's probably something I did before and forgot to revert! Should I fix it and push again?

Comment (ctb) [reply tests/test_sourmash.py:202]: no need, just for future consideration!

**Comment by keyabarve** [comment]
Is this ready to be merged? Or is there anything else that needs to be fixed?

File: src/sourmash/sourmash_args.py (hunk 3 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py]: I think it's probably fine to just say, "file '<bad stuff in here>' does not exist".

**Comment by ctb** [comment]
I think this error message is misleading --
```
                 raise ValueError(f"file '{checkfile}' listed inside the pathlist contains a badly formatted file")
```
and should be replaced with the same text as below,
```
         raise ValueError(f"pathlist file '{filename}' does not exist")    
```

**Comment by keyabarve** [comment]
Is this in the `load_pathlist_from_file` function? Should I remove the for loop then? I'm having a hard time understanding why it needs to be replaced.

**Comment by keyabarve** [comment]
@ctb This is the error I'm getting if I change the statement:
```
 assert "pathlist file '' does not exist" in "pathlist file '{'a':1}' does not exist"
E        +  where "pathlist file '{'a':1}' does not exist" = str(ValueError("pathlist file '{'a':1}' does not exist"))
E        +    where ValueError("pathlist file '{'a':1}' does not exist") = <ExceptionInfo ValueError("pathlist file '{'a':1}' does not exist") tblen=2>.value
```

**Comment by ctb** [comment]
I'm not sure how to comment, other than to say that this logic looks inherently problematic to me :)
```
        for checkfile in file_list:
            if not os.path.exists(checkfile):
                raise ValueError(f"file '{checkfile}' listed inside the pathlist contains a badly formatted file")
```
since the error message is saying "this is a badly formatted file" but the if statement is checking to see if that file _exists_... I think the useful thing to say is "this file doesn't exist".

**Comment by keyabarve** [comment]
Hmm I see. Should I do this then: `raise ValueError(f"file '{checkfile}' inside the pathlist does not exist")`

**Comment by keyabarve** [comment]
The problem that I'm facing is that it's giving me the error I mentioned above. I'm confused about how to fix that.
Is there any other way to check for a badly formatted file in the function itself?

**Comment by keyabarve** [comment]
The assert statement in my test case is this: `assert "file '' inside the pathlist does not exist" in str(e.value)`. If I change that to `assert "file '{'a':1}' inside the pathlist does not exist"` in str(e.value), then the tests pass. So should I do that then?

**Comment by ctb** [comment]
On Fri, May 21, 2021 at 02:29:36PM -0700, Keya Barve wrote:
> The assert statement in my test case is this: `assert "file '' inside the pathlist does not exist" in str(e.value)`. If I change that to `assert "file '{'a':1}' inside the pathlist does not exist"` in str(e.value), then the tests pass. So should I do that then?

Sure, give it a try.

**Comment by keyabarve** [comment]
Please review. @ctb

**Review by ctb — APPROVED** [review]
Looks good to me! I'll update from latest and then one of us can merge when all the tests pass!

Nice work - I know it took a while ;)

**Comment by keyabarve** [comment]
Alright! Thanks! By merge, you mean 'Squash and merge' right?

**Comment by keyabarve** [comment]
Ready for merge. @ctb

**Comment by ctb** [comment]
🎉
